events/listener.go

```diff
@@ -0,0 +1,675 @@
+type listenerHandle struct {
+	unsubscribe func()
+}
+
+// Stop stops the event listener
+func (h *listenerHandle) Unsubscribe() {
+	h.unsubscribe()
+}
+
+// RegisterServiceAddedHandler registers a handler for the ServiceAdded event on a CMAccount
+func (el *EventListener) RegisterServiceAddedHandler(cmAccountAddr common.Address, serviceName []string, handler EventHandler) (ListenerHandle, error) {
+	el.mu.Lock()
+	defer el.mu.Unlock()
+
+	subID := uuid.New().String()
+
+	// Get or create CMAccount instance
+	cmAccount, err := el.getOrCreateCMAccount(cmAccountAddr)
+	if err != nil {
+		return nil, err
+	}
+
+	// Create a context with cancellation
+	ctx, cancel := context.WithCancel(context.Background())
+
+	// Store subscription info
+	subInfo := &subscriptionInfo{
+		subID:       subID,
+		contract:    cmAccountAddr,
+		handler:     handler,
+		unsubscribe: cancel,
+	}
+	el.subscriptions[subID] = subInfo
+
+	// Start the resubscription
+	go el.resubscribeServiceAdded(ctx, subID, cmAccount, serviceName)
+
+	// Return handle to stop listening
+	return &listenerHandle{
+		unsubscribe: func() {
+			el.unsubscribe(subID)
+		},
+	}, nil
+}
+
+// resubscribeServiceAdded handles resubscription for ServiceAdded events
+func (el *EventListener) resubscribeServiceAdded(ctx context.Context, subID string, cmAccount *cmaccount.Cmaccount, serviceName []string) {
+	backoffMax := 2 * time.Minute // Maximum backoff time between retries
+
+	resubscribeFn := func(ctx context.Context, lastError error) (event.Subscription, error) {
+		if lastError != nil {
+			el.logger.Errorf("Resubscribe attempt after error: %v", lastError)
+		}
+
+		eventChan := make(chan *cmaccount.CmaccountServiceAdded)
+
+		// Subscribe to the event
+		sub, err := cmAccount.WatchServiceAdded(&bind.WatchOpts{Context: ctx}, eventChan, serviceName)
+		if err != nil {
+			el.logger.Errorf("Failed to subscribe to ServiceAdded events: %v", err)
+			return nil, err
+		}
+
+		el.mu.Lock()
+		subInfo, exists := el.subscriptions[subID]
+		if !exists {
+			el.mu.Unlock()
+			sub.Unsubscribe()
+			return nil, fmt.Errorf("subscription %s no longer exists", subID)
+		}
+		subInfo.sub = sub
+		subInfo.closeChan = func() {
+			close(eventChan)
+		}
+		el.mu.Unlock()
+
+		go el.listenForServiceAddedEvents(subID, eventChan)
+
+		el.logger.Infof("Subscribed for ServiceAdded events on CMAccount %s", subInfo.contract)
+		el.logger.Debugf("Subscription ID: %s", subID)
+		el.logger.Debugf("Filters: %v", serviceName)
+
+		return sub, nil
+	}
+
+	// Start resubscribing
+	sub := event.ResubscribeErr(backoffMax, resubscribeFn)
+
+	// Wait until context is canceled
+	<-ctx.Done()
+	sub.Unsubscribe()
+}
+
+// listenForServiceAddedEvents listens for ServiceAdded events
+func (el *EventListener) listenForServiceAddedEvents(subID string, eventChan chan *cmaccount.CmaccountServiceAdded) {
+	el.mu.Lock()
+	subInfo, exists := el.subscriptions[subID]
+	el.mu.Unlock()
+	if !exists {
+		return
+	}
+	handler := subInfo.handler
```

Comment:
can subInfo be accessed concurrently?

Reply:
Updated to use locks where there is a write operation.

The handler here set once during registration and does not change after. So, I think it's safe to read it without sync?